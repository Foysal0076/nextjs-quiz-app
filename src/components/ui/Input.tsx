import React, { forwardRef } from 'react'

import { cn } from '@/shared/utils'

type TextFieldParams = React.InputHTMLAttributes<HTMLInputElement> & {
  label?: string
  labelClassName?: string
  helperText?: string
  layout?: 'horizontal' | 'vertical'
  error?: string
  startAdornment?: StartAdornment
  endAdornment?: EndAdornment
}

type StartAdornment = {
  adornment: React.ReactNode
  onClick?: () => void
  className?: string
}

type EndAdornment = {
  adornment: React.ReactNode
  onClick: () => void
  className?: string
}

export const Input = forwardRef<HTMLInputElement, TextFieldParams>(
  function CustomInput(
    {
      label,
      error,
      layout = 'vertical',
      className = '',
      labelClassName = '',
      helperText,
      startAdornment,
      endAdornment,
      ...inputParams
    },
    ref
  ) {
    return (
      <div
        className={cn('custom-input group flex', {
          'flex-col gap-1': layout === 'vertical',
        })}>
        {label && (
          <label
            htmlFor={inputParams.id as string}
            className={cn(
              'text-xs font-medium',
              {
                'text-neutral-600 group-focus-within:text-primary-500 dark:group-focus-within:text-primary-300':
                  !error,
                'text-red-500 group-focus-within:text-red-500': error,
                'after:ml-0.5 after:mr-1 after:text-red-500 after:content-["*"]':
                  inputParams.required,
                'mt-2': layout === 'horizontal',
              },
              labelClassName
            )}>
            {label}
            {layout === 'horizontal' && <span>:</span>}
          </label>
        )}
        <div className='relative flex flex-col'>
          <input
            ref={ref}
            {...inputParams}
            className={cn(
              'rounded-lg border-none px-4 text-neutral-600 outline-none ring-1 placeholder:text-neutral-50 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-surface-100 dark:text-neutral-600 dark:disabled:bg-neutral-400',
              {
                'ring-neutral-50 focus:shadow-[0_0_10px_#c9bfff] focus:ring-primary-300 dark:focus:ring-primary-300':
                  !error,
                'ring-red-500 focus:ring-red-500': error,
                'pr-12': endAdornment,
                'pl-10 max-xs:max-w-[13.5rem]': startAdornment,
              },
              className
            )}
          />
          {helperText && (
            <span className='helper-text mt-1 text-xs font-normal text-neutral-400'>
              {helperText}
            </span>
          )}
          {error && (
            <div
              className={cn(
                'absolute -bottom-0 left-0',
                helperText ? 'translate-y-[80%]' : 'translate-y-[90%]'
              )}>
              <span className='text-xs text-red-500 dark:text-red-500'>
                {error}
              </span>
            </div>
          )}
          {startAdornment && (
            <button
              type='button'
              onClick={startAdornment?.onClick}
              className={cn(
                'absolute left-2 top-1/2 w-max -translate-y-1/2 rounded-full text-neutral-50 transition-colors duration-200 hover:bg-neutral-30',
                { 'pointer-events-none': !startAdornment.onClick },
                startAdornment.className
              )}
              aria-label='start-adornment btn'>
              {startAdornment.adornment}
            </button>
          )}
          {endAdornment && (
            <button
              type='button'
              onClick={endAdornment?.onClick}
              className={cn(
                'absolute right-2 top-1/2 w-max -translate-y-1/2 rounded-full text-neutral-50 transition-colors duration-200 hover:bg-neutral-30',
                { 'pointer-events-none': !endAdornment.onClick },
                endAdornment.className
              )}
              aria-label='end-adornment btn'>
              {endAdornment.adornment}
            </button>
          )}
        </div>
      </div>
    )
  }
)
