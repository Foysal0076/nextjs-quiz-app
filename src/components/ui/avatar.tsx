import Image from 'next/image'

import { cn, createInitial } from '@/shared/utils'

type Props = {
  name: string
  url?: string
  size?: number
  className?: string
  textClassName?: string
}

const Avatar = ({ name, url, className, textClassName }: Props) => {
  const initial = createInitial(name)

  return (
    <div
      className={cn(
        `relative flex h-8 w-8 items-center justify-center rounded-full border-neutral-100 bg-surface-300 text-base font-bold`,
        className
      )}>
      {url ? (
        <Image
          src={url}
          alt={name}
          fill
          className='rounded-full object-cover'
          sizes='(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw'
          placeholder='blur'
          blurDataURL={`_next/image?url=${url}&w=16&q=1`}
        />
      ) : (
        <span className={cn('text-sm font-bold md:text-base', textClassName)}>
          {initial}
        </span>
      )}
    </div>
  )
}

export default Avatar
