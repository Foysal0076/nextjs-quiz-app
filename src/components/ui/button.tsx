import { cva } from 'class-variance-authority'

import Spinner from '@/components/ui/spinner'
import { cn } from '@/shared/utils'

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: 'primary' | 'secondary' | 'danger' | 'success'
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
  fullWidth?: boolean
  loading?: boolean
}

const ButtonVariants = cva(
  'rounded-lg transition-all duration-200 text-nowrap focus:outline-none font-medium',
  {
    variants: {
      variant: {
        primary:
          'bg-primary-500 text-white hover:bg-primary-600 focus:ring-4 focus:ring-primary-100 text-neutral-0 focus:outline-primary-200 disabled:active:transform-none active:scale-[.95] disabled:text-neutral-200 disabled:bg-primary-100 disabled:cursor-not-allowed text-neutral-0 dark:text-neutral-900',
        secondary:
          'bg-none text-neutral-900 hover:bg-primary-50 focus:ring-primary-100 border-2 border-primary-500 text-primary-500 focus:ring-4 focus:ring-primary-100 disabled:active:transform-none active:scale-[.95] disabled:cursor-not-allowed disabled:text-neutral-200 disabled:border-neutral-50 disabled:hover:bg-neutral-0 dark:text-primary-200 dark:border-primary-200 dark:hover:bg-surface-200 dark:disabled:border-neutral-200 dark:disabled:text-neutral-200',
        danger:
          'bg-red-500 hover:bg-red-600 text-neutral-0 focus:ring-4 focus:ring-red-100 disabled:active:transform-none active:scale-[.95] disabled:text-neutral-200 disabled:bg-neutral-50 disabled:cursor-not-allowed text-neutral-0 dark:text-neutral-900',
        success:
          'bg-green-500 hover:bg-green-600 text-neutral-0 focus:ring-4 focus:ring-green-100 disabled:active:transform-none active:scale-[.95] disabled:text-neutral-200 disabled:bg-neutral-50 disabled:cursor-not-allowed text-neutral-0 dark:text-neutral-900',
      },
      size: {
        xs: 'h-[1.75rem] text-xs px-4',
        sm: 'h-[2rem] text-sm px-6',
        md: 'h-[2.5rem] text-md px-8',
        lg: 'h-[3rem] text-md px-11',
        xl: 'h-[3.5rem] text-lg px-11',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)

const Button = ({
  variant = 'primary',
  size = 'md',
  loading = false,
  className = '',
  children,
  ...buttonProps
}: ButtonProps) => {
  return (
    <button
      className={cn(
        `relative select-none`,
        ButtonVariants({ variant, size }),
        className
      )}
      disabled={loading}
      {...buttonProps}>
      {loading && (
        <div
          className='absolute inset-0 flex items-center justify-center'
          data-testid='spinner'>
          <Spinner />
        </div>
      )}
      <div
        className={cn(`visible`, {
          invisible: loading,
          'opacity-0': loading,
        })}>
        {children}
      </div>
    </button>
  )
}

export default Button
