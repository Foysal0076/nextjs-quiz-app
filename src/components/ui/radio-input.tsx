import { forwardRef, InputHTMLAttributes } from 'react'

import { cn } from '@/shared/utils'

type RadioProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'type'> & {
  label?: string
}

const Radio = forwardRef<HTMLInputElement, RadioProps>(
  ({ className, label, id, ...props }, ref) => {
    return (
      <div className='flex items-center space-x-2'>
        <input
          {...props}
          ref={ref}
          type='radio'
          id={id}
          className={cn(
            'h-4 w-4 border-gray-300 bg-gray-100 text-primary-600 focus:ring-2 focus:ring-primary-500 dark:border-gray-600 dark:bg-gray-700 dark:checked:bg-primary-500 dark:focus:ring-primary-600',
            className
          )}
        />
        {label && (
          <label
            htmlFor={id}
            className='cursor-pointer select-none text-sm font-medium text-neutral-700'>
            {label}
          </label>
        )}
      </div>
    )
  }
)

export default Radio
