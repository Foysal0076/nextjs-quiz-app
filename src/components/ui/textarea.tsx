import React, { forwardRef } from 'react'

import { cn } from '@/shared/utils'

type TextAreaParams = React.TextareaHTMLAttributes<HTMLTextAreaElement> & {
  label?: string
  labelClassName?: string
  helperText?: string
  layout?: 'horizontal' | 'vertical'
  error?: string
}

export const Textarea = forwardRef<HTMLTextAreaElement, TextAreaParams>(
  function CustomTextarea(
    {
      label,
      error,
      layout = 'vertical',
      className = '',
      labelClassName = '',
      helperText,
      ...textareaParams
    },
    ref
  ) {
    return (
      <div
        className={cn('custom-textarea group flex', {
          'flex-col gap-1': layout === 'vertical',
        })}>
        {label && (
          <label
            htmlFor={textareaParams.id as string}
            className={cn(
              'text-sm font-semibold',
              {
                'text-neutral-600 group-focus-within:text-primary-500 dark:group-focus-within:text-primary-300':
                  !error,
                'text-red-500 group-focus-within:text-red-500': error,
                'after:ml-0.5 after:mr-1 after:text-red-500 after:content-["*"]':
                  textareaParams.required,
                'mt-2': layout === 'horizontal',
              },
              labelClassName
            )}>
            {label} {layout === 'horizontal' && <span>:</span>}
          </label>
        )}
        <div className='relative flex flex-col'>
          <textarea
            ref={ref}
            {...textareaParams}
            className={cn(
              'rounded-lg border-none px-4 py-2 text-neutral-600 outline-none ring-1 placeholder:text-neutral-50 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-surface-100 dark:text-neutral-600 dark:disabled:bg-neutral-400',
              {
                'ring-neutral-50 focus:shadow-[0_0_10px_#c9bfff] focus:ring-primary-300 dark:focus:ring-primary-300':
                  !error,
                'ring-red-500 focus:ring-red-500': error,
              },
              className
            )}
          />
          {helperText && (
            <span className='helper-text mt-1 text-xs font-normal text-neutral-400'>
              {helperText}
            </span>
          )}
          {error && (
            <div
              className={cn(
                'absolute -bottom-0 left-0',
                helperText ? 'translate-y-[80%]' : 'translate-y-[90%]'
              )}>
              <span className='text-xs text-red-500 dark:text-red-500'>
                {error}
              </span>
            </div>
          )}
        </div>
      </div>
    )
  }
)
