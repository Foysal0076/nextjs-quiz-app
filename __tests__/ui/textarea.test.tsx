import { Textarea } from '@/components/ui/textarea'
import { render, screen } from '@testing-library/react'

describe('Textarea Component', () => {
  it('renders textarea correctly', () => {
    render(<Textarea placeholder='Enter text' />)
    expect(screen.getByPlaceholderText('Enter text')).toBeInTheDocument()
  })

  it('renders label when provided', () => {
    render(<Textarea label='Description' id='desc' />)
    expect(screen.getByText('Description')).toBeInTheDocument()
  })

  it('displays error message when error prop is provided', () => {
    const errorMessage = 'This field is required'
    render(<Textarea error={errorMessage} />)
    expect(screen.getByText(errorMessage)).toBeInTheDocument()
  })

  it('displays helper text when provided', () => {
    const helperText = 'Maximum 500 characters'
    render(<Textarea helperText={helperText} />)
    expect(screen.getByText(helperText)).toBeInTheDocument()
  })

  it('applies horizontal layout correctly', () => {
    render(<Textarea label='Notes' layout='horizontal' />)
    const label = screen.getByText('Notes')
    expect(label).toBeInTheDocument()
  })

  it('shows required asterisk when textarea is required', () => {
    render(<Textarea label='Required Field' required />)
    const labelContainer = screen.getByText('Required Field')
    expect(labelContainer).toHaveClass('after:content-["*"]')
  })

  it('applies custom className to textarea', () => {
    const customClass = 'custom-textarea'
    render(<Textarea className={customClass} />)
    const textarea = screen.getByRole('textbox')
    expect(textarea).toHaveClass(customClass)
  })

  it('applies custom labelClassName to label', () => {
    const customLabelClass = 'custom-label'
    render(<Textarea label='Custom Label' labelClassName={customLabelClass} />)
    const label = screen.getByText('Custom Label')
    expect(label).toHaveClass(customLabelClass)
  })

  it('applies error styles when error is provided', () => {
    render(<Textarea error='Error message' />)
    const textarea = screen.getByRole('textbox')
    expect(textarea).toHaveClass('ring-red-500')
  })

  it('forwards ref correctly', () => {
    const ref = jest.fn()
    render(<Textarea ref={ref} />)
    expect(ref).toHaveBeenCalled()
  })

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(
      <Textarea
        label='Description'
        helperText='Enter your description'
        placeholder='Type here...'
        required
      />
    )
    expect(container).toMatchSnapshot()
  })
})
