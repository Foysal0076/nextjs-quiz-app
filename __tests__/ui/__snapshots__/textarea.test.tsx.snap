// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Component matches snapshot 1`] = `
<div>
  <div
    class="custom-textarea group flex flex-col gap-1"
  >
    <label
      class="text-sm font-semibold text-neutral-600 group-focus-within:text-primary-500 dark:group-focus-within:text-primary-300 after:ml-0.5 after:mr-1 after:text-red-500 after:content-["*"]"
    >
      Description
       
    </label>
    <div
      class="relative flex flex-col"
    >
      <textarea
        class="rounded-lg border-none px-4 py-2 text-neutral-600 outline-none ring-1 placeholder:text-neutral-50 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-surface-100 dark:text-neutral-600 dark:disabled:bg-neutral-400 ring-neutral-50 focus:shadow-[0_0_10px_#c9bfff] focus:ring-primary-300 dark:focus:ring-primary-300"
        placeholder="Type here..."
        required=""
      />
      <span
        class="helper-text mt-1 text-xs font-normal text-neutral-400"
      >
        Enter your description
      </span>
    </div>
  </div>
</div>
`;
